pytorch_lightning/plugins/training_type/ipu.py

```diff
@@ -0,0 +1,236 @@
+import inspect
+import json
+import os
+from typing import Any, Iterable, List, Optional, Union
+
+import torch
+from torch.nn import Module
+from torch.utils.data import DataLoader
+
+from pytorch_lightning import _logger as log
+from pytorch_lightning.core.lightning import LightningModule
+from pytorch_lightning.overrides.base import _LightningModuleWrapperBase
+from pytorch_lightning.plugins.environments.cluster_environment import ClusterEnvironment
+from pytorch_lightning.plugins.training_type.parallel import ParallelPlugin
+from pytorch_lightning.trainer.supporters import CombinedLoader
+from pytorch_lightning.utilities import _POPTORCH_AVAILABLE
+from pytorch_lightning.utilities.apply_func import apply_to_collection
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+if _POPTORCH_AVAILABLE:
+    import poptorch
+
+# todo: Check gradient accumulation to ensure this works, similar to DeepSpeed IPUs manage this.
+# todo: Check lr scheduling to ensure that when the LR is changed, we update the optimizer state.
+
+
+class LightningIPUModule(_LightningModuleWrapperBase):
+
+    def __init__(self, pl_module: LightningModule, precision: int):
+        super().__init__(pl_module)
+        self.precision = precision
+
+    def forward(self, *inputs, **kwargs):
+        if self.precision == 16:
+            inputs = self._move_float_tensors_to_half(inputs)
+
+        return super().forward(*inputs, **kwargs)
+
+    @staticmethod
+    def batch_to(data):
+        return data.half()
+
+    def _move_float_tensors_to_half(self, batch: Any):
+        batch = apply_to_collection(batch, (torch.FloatTensor, torch.cuda.FloatTensor), function=self.batch_to)
+        return batch
+
+
+class IPUPlugin(ParallelPlugin):
+
+    def __init__(
+        self,
+        half: bool = False,
+        device_iterations: int = 1,
```

Comment:
These two options are important. `half` should be changed to a better name, but allows pure FP16 training (where the model weights are in FP16). `device_iterations` is the number of iterations to run on the device before returning (imagine running 100 batches of data on the device then returning).

it would be nice to expose them to the trainer, however this might look like overcrowding (`ipu_half` or `ipu_device_iterations`).

Because we rely on an accelerator it makes the API a bit tricky when instantiating the object:

```python
trainer = Trainer(
        accelerator=IPUAccelerator(
            training_type_plugin=IPUPlugin(device_iterations=32),
            precision_plugin=IPUPrecisionPlugin(precision=16)
        ),
        precision=16
    )
```

All that is really needed is the IPUPlugin, but I'm not sure if there is logic in place to auto-select the IPU Accelerator if we do so. then it could become:

```python
trainer = Trainer(
        plugin=IPUPlugin(device_iterations=32),
        precision=16
    )
```

Reply:
That logic should be easy to add I think. But that could also be part of #6090 .

One more reason to introduce something like `num_devices` and `device_type` instead of having flags for every device separately.